go/consensus/tendermint/apps/registry/transactions.go

```diff
@@ -241,12 +242,71 @@ func (app *registryApplication) registerNode( // nolint: gocyclo
 
 	// Check runtime's whitelist.
 	for _, rt := range paidRuntimes {
-		if rt.AdmissionPolicy.EntityWhitelist != nil && !rt.AdmissionPolicy.EntityWhitelist.Entities[newNode.EntityID] {
-			ctx.Logger().Error("RegisterNode: node's entity not in a runtime's whitelist",
-				"entity", newNode.EntityID,
-				"runtime", rt.ID,
-			)
-			return registry.ErrForbidden
+		if rt.AdmissionPolicy.EntityWhitelist != nil {
+			wcfg, entIsWhitelisted := rt.AdmissionPolicy.EntityWhitelist.Entities[newNode.EntityID]
+			if !entIsWhitelisted {
+				ctx.Logger().Error("RegisterNode: node's entity not in a runtime's whitelist",
+					"entity", newNode.EntityID,
+					"runtime", rt.ID,
+				)
+				return registry.ErrForbidden
+			}
+			if wcfg.MaxNodes != nil && len(wcfg.MaxNodes) > 0 {
+				// Map is present and non-empty, check per-role restrictions
+				// on the maximum number of nodes per entity.
+
+				// Iterate over all valid roles (each entry in the map can
+				// only have a single role).
+				for _, role := range node.Roles() {
+					if !newNode.HasRoles(role) {
+						// Skip unset roles.
+						continue
+					}
+
+					maxNodes, exists := wcfg.MaxNodes[role]
+					if !exists {
+						// No such role found in whitelist.
+						return registry.ErrForbidden
+					}
+					if maxNodes == 0 {
+						// No nodes of this type are allowed.
+						return registry.ErrForbidden
+					}
+
+					// Count existing nodes owned by entity.
+					ent, grr := state.Entity(ctx, newNode.EntityID)
```

Comment:
Instead of querying the entity descriptor you should query the `nodeByEntityKeyFmt` (add state accessor, e.g. `GetEntityNodes`, as currently there is only `HasEntityNodes`). Also expired nodes (`n.IsExpired(epoch)`) should not count against the maximum as those are not eligible to be scheduled.